language/tools/move-cli/src/package.rs

```diff
@@ -0,0 +1,227 @@
+                .path()
+                .join(self.name)
+                .join(PKG_SOURCE_DIR)
+                .into_os_string()
+                .into_string()
+                .unwrap();
+            // need to exclude self from deps
+            deps.retain(|path| path != &tmp_src_dir);
+
+            // compile
+            let (_, compiled_units) = move_compile(&[tmp_src_dir], &deps, None, None)?;
+
+            // save compiled units to files
+            for unit in compiled_units {
+                let mut data = vec![];
+                let name = match unit {
+                    CompiledUnit::Module { ident, module, .. } => {
+                        module.serialize(&mut data)?;
+                        ident.0.value.name.0.value
+                    }
+                    CompiledUnit::Script { key, script, .. } => {
+                        script.serialize(&mut data)?;
+                        key
+                    }
+                };
+                let file_path = pkg_path.join(name).with_extension(COMPILED_EXTENSION);
+                let mut fp = File::create(file_path)?;
+                fp.write_all(&data)?;
+            }
+        }
+
+        // done
+        Ok(dep_dirs)
+    }
+
+    /// Load modules from the given directory
+    pub fn load_modules(&self, dir_path: &Path) -> Result<Vec<CompiledModule>> {
+        let compiled_modules =
+            find_filenames(
+                &[dir_path.to_str().unwrap().to_owned()],
+                |fpath| match fpath.extension().and_then(|s| s.to_str()) {
+                    None => false,
+                    Some(ext) => ext == COMPILED_EXTENSION,
+                },
+            )?
+            .iter()
+            .filter_map(|entry| {
+                CompiledModule::deserialize(&fs::read(Path::new(entry)).unwrap()).ok()
+            })
+            .collect();
+        Ok(compiled_modules)
+    }
+}
+
+/// The dependency interface exposed to CLI main
+pub struct MoveDep(Vec<&'static MovePackage>);
+
+impl MoveDep {
+    pub fn prepare_source_files(&self, dir_path: &Path) -> Result<Vec<String>> {
+        let mut paths = vec![];
+        for pkg in &self.0 {
+            paths.extend(pkg.prepare_source_files(dir_path)?);
+        }
+        Ok(paths)
+    }
+
+    pub fn prepare_binary_files(&self, dir_path: &Path) -> Result<Vec<String>> {
+        let mut paths = vec![];
+        for pkg in &self.0 {
+            paths.extend(pkg.prepare_binary_files(dir_path)?);
+        }
+        Ok(paths)
+    }
+
+    pub fn load_modules(&self, dir_path: &Path) -> Result<Vec<CompiledModule>> {
+        let mut modules = vec![];
+        for pkg in &self.0 {
+            modules.extend(pkg.load_modules(dir_path)?);
+        }
+        Ok(modules)
+    }
+}
+
+pub fn parse_deps_from_mode_string(mode: &str) -> Result<MoveDep> {
+    match mode {
+        "bare" => Ok(MoveDep(vec![])),
+        "stdlib" => Ok(MoveDep(vec![&PACKAGE_NURSERY])),
```

Comment:
I don't think this is quite the split we want. 

There should be three modes 
"bare" -> empty as it is here today
"stdlib" -> [Vector, Option, Errors, Events, ..Nursery]
"libra" -> [/* Everything else */, ..Stdlib]

Reply:
Thank you! This is precisely the things I would like to ask you and Tim. I also noticed that there is a bundle of transaction scripts sitting in stdlib, should they belong to `libra` too? Or we don't need to worry about them?

Reply:
I don't think you need to worry about the transaction scripts. They're predefined since custom scripts aren't allowed to be sent to the chain, but for the CLI they are.

Reply:
Thanks for the clarification Tim.

Reply:
So, I plan to put everything in the [Module Utility Libraries](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/overview.md#module-utility-libraries) plus nursery as `stdlib` and the rest as `libra`.

Right now I guess I have to hardcode this information though...

Reply:
I think we will always have to hard code it unless we split up or directories. Some work for another time @tzakian

Reply:
Agreed. I think for now this will need to be hardcoded, but eventually we'll want to split this up.